src/lib/party.ts

```diff
@@ -494,6 +498,50 @@ type UnfinishedSignature = undefined | LazySignature | string;
 
 type LazyControl = Control | LazySignature | LazyProof;
 
+class Token {
+  readonly id: Field;
+  readonly parentTokenId: Field;
+  readonly tokenOwner: PublicKey;
+
+  constructor(options: { tokenOwner: PublicKey; parentTokenId?: Field }) {
+    let { tokenOwner, parentTokenId } = options ?? {};
+
+    // Reassign to default tokenId if undefined
+    parentTokenId ??= getDefaultTokenId();
+
+    // Check if we can create a custom tokenId
+    try {
+      Ledger.customTokenId(tokenOwner, parentTokenId);
+    } catch (e) {
+      throw new Error(
+        `Could not create a custom token id:\nError: ${(e as Error).message}`
+      );
+    }
+
+    this.parentTokenId = parentTokenId;
+    this.tokenOwner = tokenOwner;
+    if (
+      tokenOwner.toFields().every((x) => x.isConstant()) &&
+      parentTokenId.isConstant()
+    ) {
+      this.id = Ledger.customTokenIdChecked(tokenOwner, this.parentTokenId);
```

Comment:
What's the difference between these branches?

Reply:
hey @MartinMinkov, I think these branches are wrong.. the else branch should use the checked version, not the if branch

@bkase, if the token owner or parent token id is a variable then we need to derive the token id in-snark. (Not sure if we could always use the checked version, depends on how flexible it is written, but IIRC checked hashing returns a new variable even when the inputs are constant, which wouldn't work outside the circuit)

Reply:
Yikes! Good catch, switch these. :sweat_smile: